plugins/googleanalytics/src/org/pentaho/di/trans/steps/googleanalytics/GaInputStep.java

```diff
@@ -0,0 +1,266 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ ******************************************************************************/
+
+package org.pentaho.di.trans.steps.googleanalytics;
+
+import java.io.IOException;
+import java.util.List;
+
+import com.google.api.services.analytics.Analytics;
+import org.pentaho.di.core.Const;
+import org.pentaho.di.core.exception.KettleException;
+import org.pentaho.di.core.row.RowDataUtil;
+import org.pentaho.di.core.row.RowMeta;
+import org.pentaho.di.core.row.ValueMetaInterface;
+import org.pentaho.di.core.row.value.ValueMetaFactory;
+import org.pentaho.di.i18n.BaseMessages;
+import org.pentaho.di.trans.Trans;
+import org.pentaho.di.trans.TransMeta;
+import org.pentaho.di.trans.step.BaseStep;
+import org.pentaho.di.trans.step.StepDataInterface;
+import org.pentaho.di.trans.step.StepInterface;
+import org.pentaho.di.trans.step.StepMeta;
+import org.pentaho.di.trans.step.StepMetaInterface;
+
+
+public class GaInputStep extends BaseStep implements StepInterface {
+
+  private static Class<?> PKG = GaInputStepMeta.class; // for i18n purposes
+
+  private GaInputStepData data;
+  private GaInputStepMeta meta;
+
+  private Analytics analytics;
+  private String accountName;
+
+  public GaInputStep( StepMeta s, StepDataInterface stepDataInterface, int c, TransMeta t, Trans dis ) {
+    super( s, stepDataInterface, c, t, dis );
+  }
+
+  public boolean processRow( StepMetaInterface smi, StepDataInterface sdi ) throws KettleException {
+
+    meta = (GaInputStepMeta) smi;
+    data = (GaInputStepData) sdi;
+
+    if ( first ) {
+
+      first = false;
+
+      data.outputRowMeta = new RowMeta();
+      meta.getFields( data.outputRowMeta, getStepname(), null, null, this, repository, metaStore );
+
+      // stores the indices where to look for the key fields in the input rows
+      data.conversionMeta = new ValueMetaInterface[meta.getFeedField().length];
+
+      for ( int i = 0; i < meta.getFeedField().length; i++ ) {
+
+        // get output and from-string conversion format for each field
+        ValueMetaInterface returnMeta = data.outputRowMeta.getValueMeta( i );
+
+        ValueMetaInterface conversionMeta;
+
+        conversionMeta = ValueMetaFactory.cloneValueMeta( returnMeta, ValueMetaInterface.TYPE_STRING );
+        conversionMeta.setConversionMask( meta.getConversionMask()[i] );
+        conversionMeta.setDecimalSymbol( "." ); // google analytics is en-US
+        conversionMeta.setGroupingSymbol( null ); // google analytics uses no grouping symbol
+
+        data.conversionMeta[i] = conversionMeta;
+      }
+    }
+
+    // generate output row, make it correct size
+    Object[] outputRow = RowDataUtil.allocateRowData( data.outputRowMeta.size() );
+
+    List<String> entry = null;
+
+    // no query prepared yet?
+    if ( data.query == null ) {
```

Comment:
Any reason this code was moved from the getNextEntry method? Seems like having it all crammed into the processRow makes processRow less readable.